coix/numpyro.py

```diff
@@ -42,11 +45,18 @@ def wrapped(*args, **kwargs):
         trace[name] = {"value": value, "log_prob": log_prob}
         if site.get("is_observed", False):
           trace[name]["is_observed"] = True
+    # else:
+    # in_log_weight = _get_log_weight(q_trace, batch_ndims)
     metrics = {
         name: site["value"]
         for name, site in tr.items()
         if site["type"] == "metric"
     }
+    # add log_weight to metrics
+    if "log_weight" not in metrics:
+      log_probs = [get_site_log_prob(site) for site in trace.values()]
+      weight = _get_log_weight(trace, _get_batch_ndims(log_probs))
+      metrics = {**metrics, "log_weight": weight}
```

Comment:
after this, I think we can add a note to `compose` that currently, if one of q1 or q2 is an inference program, then the other is a primitive without observed variables (a kernel?).

Reply:
Yes, I call them kernel programs in the tutorials now. I think that's an appropriate name.